### ISSUE TITLE
My model does not converge

### ISSUE THREAD

**Issue by kyu899** [Microsoft/CNTK#1143]
Hi.
I tried to train my own, very small and simple CNN by modifying the CifarResnet.cpp located in Tests\UnitTests\V2LibraryTests

But the model seems to be not trained at all.
For instance, it prints the results by following:
Minibatch 0: CrossEntropy loss = 2.2925203, Evaluation criterion = 0.875
Minibatch 60: CrossEntropy loss = 2.300297, Evaluation criterion = 0.890625
Minibatch 120: CrossEntropy loss = 2.3074899, Evaluation criterion = 0.921875
Minibatch 180: CrossEntropy loss = 2.3153296, Evaluation criterion = 0.875
until Minibatch 50,000 the error do not decrease.

I changed learning rate, use different SGD schemes, but they do not solve the problem.

Please refer to my source code and let me know what do I wrong.

Thank you so much.

#include "CNTKLibrary.h"
#include <functional>
#include "Common.h"
#include "Image.h"

using namespace CNTK;
using namespace std;
using namespace std::placeholders;

MinibatchSourcePtr CreateCifarMinibatchSource(size_t epochSize)
{
    size_t imageHeight = 32;
    size_t imageWidth = 32;
    size_t numChannels = 3;
    size_t numClasses = 10;

	fprintf(stderr, "Load Image files...\n");
    auto mapFilePath = L"train_map.txt";
    auto meanFilePath = L"CIFAR-10_mean.xml";

    Dictionary cropTransformConfig;
    cropTransformConfig[L"type"] = L"Crop";
    cropTransformConfig[L"cropType"] = L"Random";
    cropTransformConfig[L"cropRatio"] = L"0.8";
    cropTransformConfig[L"jitterType"] = L"uniRatio";

    Dictionary scaleTransformConfig;
    scaleTransformConfig[L"type"] = L"Scale";
    scaleTransformConfig[L"width"] = imageWidth;
    scaleTransformConfig[L"height"] = imageHeight;
    scaleTransformConfig[L"channels"] = numChannels;
    scaleTransformConfig[L"interpolations"] = L"linear";

    Dictionary meanTransformConfig;
    meanTransformConfig[L"type"] = L"Mean";
    meanTransformConfig[L"meanFile"] = meanFilePath;

    std::vector<DictionaryValue> allTransforms = { cropTransformConfig, scaleTransformConfig, meanTransformConfig };

    Dictionary featuresStreamConfig;
    featuresStreamConfig[L"transforms"] = allTransforms;

    Dictionary labelsStreamConfig;
    labelsStreamConfig[L"labelDim"] = numClasses;

    Dictionary inputStreamsConfig;
    inputStreamsConfig[L"features"] = featuresStreamConfig;
    inputStreamsConfig[L"labels"] = labelsStreamConfig;

    Dictionary deserializerConfiguration;
    deserializerConfiguration[L"type"] = L"ImageDeserializer";
    deserializerConfiguration[L"file"] = mapFilePath;
    deserializerConfiguration[L"input"] = inputStreamsConfig;

    Dictionary minibatchSourceConfiguration;
    minibatchSourceConfiguration[L"epochSize"] = epochSize;
    minibatchSourceConfiguration[L"deserializers"] = std::vector<DictionaryValue>({ deserializerConfiguration });

	fprintf(stderr, "Create minibatch sources...\n");
    return CreateCompositeMinibatchSource(minibatchSourceConfiguration);
}

Constant GetProjectionMap(size_t outputDim, size_t inputDim, const DeviceDescriptor& device)
{
    if (inputDim > outputDim)
        throw std::runtime_error("Can only project from lower to higher dimensionality");

    std::vector<float> projectionMapValues(inputDim * outputDim, 0);
    for (size_t i = 0; i < inputDim; ++i)
        projectionMapValues[(i * inputDim) + i] = 1.0f;

    auto projectionMap = MakeSharedObject<NDArrayView>(DataType::Float, NDShape({ 1, 1, inputDim, outputDim }), device);
    projectionMap->CopyFrom(NDArrayView(NDShape({ 1, 1, inputDim, outputDim }), projectionMapValues));

    return Constant(projectionMap);
}

FunctionPtr MyCNN(Variable input, size_t numOutputClasses, const DeviceDescriptor& device, const std::wstring& outputName)
{
	size_t kernelWidth = 5;
	size_t kernelHeight = 5;

	fprintf(stderr, "numInputChannels...\n");
	size_t numInputChannels = input.Shape()[input.Shape().Rank() - 1];

	auto conv1Params = Parameter(NDArrayView::RandomUniform<float>({ kernelWidth, kernelHeight, numInputChannels, 4 }, -0.05, 0.05, 1, device), L"C1");
	auto conv1BiasParams = Parameter(NDArrayView::RandomUniform<float>({ 1, 1, 4 }, -0.05, 0.05, 1, device));
	auto c1 = ReLU(Plus(conv1BiasParams, Convolution(conv1Params, input, { 1, 1, numInputChannels })));
	auto m1 = Pooling(c1, CNTK::PoolingType::Max, { 3, 3, 1}, { 2, 2, 1});

	auto conv2Params = Parameter(NDArrayView::RandomUniform<float>({ kernelWidth, kernelHeight, 4, 4 }, -0.05, 0.05, 1, device), L"C2");
	auto conv2BiasParams = Parameter(NDArrayView::RandomUniform<float>({ 1, 1, 4 }, -0.05, 0.05, 1, device));
	auto c2 = ReLU(Plus(conv2BiasParams, Convolution(conv2Params, m1, { 1, 1, 4 })));
	auto m2 = Pooling(c2, CNTK::PoolingType::Max, { 3, 3, 1 }, { 2, 2, 1 });

	Variable V1 = m2;
	
	auto f1TimesParams = Parameter(NDArrayView::RandomUniform<float>({ 16, V1.Shape()[0], V1.Shape()[1], V1.Shape()[2] }, -0.05, 0.05, 1, device));
	auto f1BiasParams = Parameter(NDArrayView::RandomUniform<float>({ 16 }, -0.05, 0.05, 1, device));
	auto f1 = ReLU(Plus(f1BiasParams, Times(f1TimesParams, m2)));

	auto f2TimesParams = Parameter(NDArrayView::RandomUniform<float>({ numOutputClasses, 16 }, -0.05, 0.05, 1, device));
	auto f2BiasParams = Parameter(NDArrayView::RandomUniform<float>({ numOutputClasses }, -0.05, 0.05, 1, device));
	return Plus(f2BiasParams, Times(f2TimesParams, f1));
}

void TrainResNetCifarClassifer(const DeviceDescriptor& device, bool testSaveAndReLoad)
{
    auto minibatchSource = CreateCifarMinibatchSource(SIZE_MAX);
    auto imageStreamInfo = minibatchSource->StreamInfo(L"features");
    auto labelStreamInfo = minibatchSource->StreamInfo(L"labels");

    auto inputImageShape = imageStreamInfo.m_sampleLayout;
    const size_t numOutputClasses = labelStreamInfo.m_sampleLayout[0];

    auto imageInputName = L"Images";
    auto imageInput = InputVariable(inputImageShape, imageStreamInfo.m_elementType, imageInputName);
    auto classifierOutput = MyCNN(imageInput, numOutputClasses, device, L"classifierOutput");

    auto labelsInputName = L"Labels";
    auto labelsVar = InputVariable({ numOutputClasses }, labelStreamInfo.m_elementType, labelsInputName);
	auto trainingLoss = CNTK::CrossEntropyWithSoftmax(classifierOutput, labelsVar, L"lossFunction");
	auto prediction = CNTK::ClassificationError(classifierOutput, labelsVar, L"classificationError");

	LearningRatePerSampleSchedule learningRatePerSample =0.007125;
    Trainer trainer(classifierOutput, trainingLoss, prediction, { SGDLearner(classifierOutput->Parameters(), learningRatePerSample) });

    const size_t minibatchSize = 64;
    size_t numMinibatchesToTrain = 50000;
    size_t outputFrequencyInMinibatches = 60;

    for (size_t i = 0; i < numMinibatchesToTrain; ++i)
    {
	auto minibatchData = minibatchSource->GetNextMinibatch(minibatchSize, device);
        trainer.TrainMinibatch({ { imageInput, minibatchData[imageStreamInfo].m_data }, { labelsVar, minibatchData[labelStreamInfo].m_data } }, device);
        PrintTrainingProgress(trainer, i, outputFrequencyInMinibatches);
    }
}

void TrainCifarResnet()
{
	fprintf(stderr, "\nTrainCifarResnet..\n");

	if (IsGPUAvailable())
	{
		fprintf(stderr, "test on a NVIDIA GPU device.\n");
		TrainResNetCifarClassifer(DeviceDescriptor::GPUDevice(0), false /*testSaveAndReLoad*/);
	}
	else
	{
		fprintf(stderr, "test on a CPU device.\n");
		TrainResNetCifarClassifer(DeviceDescriptor::CPUDevice(), false /*testSaveAndReLoad*/);
	}
}

**Comment by cha-zhang** [comment]
Any reason you are not using Python for this task? We have examples here: 
https://github.com/Microsoft/CNTK/tree/master/Examples/Image/Classification/ConvNet

**Comment by eldakms** [comment]
Could it be your missing Transpose transform?(It is added on python by default)
Please try out the latest master/v2. Thanks!
Closing due to inactivity, feel free to reopen if needed.
